cmd/oci-catalog/src/providers/dockerhub.rs

```diff
@@ -105,25 +121,79 @@ impl OCICatalogSender for DockerHubAPI {
         }
     }
 
-    async fn send_tags(&self, tx: mpsc::Sender<Result<Tag, Status>>, _request: &ListTagsRequest) {
-        for count in 0..10 {
-            tx.send(Ok(Tag {
-                name: format!("tag-{}", count),
-            }))
-            .await
-            .unwrap();
+    async fn send_tags(&self, tx: mpsc::Sender<Result<Tag, Status>>, request: &ListTagsRequest) {
+        let mut url = match url_for_request_tags(request) {
+            Ok(u) => u,
+            Err(e) => {
+                tx.send(Err(e)).await.unwrap();
+                return;
+            }
+        };
+
+        let client = reqwest::Client::builder().build().unwrap();
+
+        loop {
+            log::debug!("requesting: {}", url);
+            let response = match client.get(url.clone()).send().await {
+                Ok(r) => r,
+                Err(e) => {
+                    tx.send(Err(Status::failed_precondition(e.to_string())))
+                        .await
+                        .unwrap();
+                    return;
+                }
+            };
+
+            if response.status() != StatusCode::OK {
+                tx.send(Err(Status::failed_precondition(format!(
+                    "unexpected status code when requesting {}: {}",
+                    url,
+                    response.status()
+                ))))
+                .await
+                .unwrap();
+                return;
+            }
+
+            let body = match response.text().await {
+                Ok(b) => b,
+                Err(e) => {
+                    tx.send(Err(Status::failed_precondition(format!(
+                        "unable to extract body from response: {}",
+                        e.to_string()
+                    ))))
+                    .await
+                    .unwrap();
+                    return;
+                }
+            };
+            log::trace!("response body: {}", body);
+
+            let response: DockerHubV2TagsResult = serde_json::from_str(&body).unwrap();
+
+            for tag in response.results {
+                tx.send(Ok(Tag { name: tag.name })).await.unwrap();
+            }
+
+            if response.next.is_some() {
+                url = reqwest::Url::parse(&response.next.unwrap()).unwrap();
```

Comment:
Ah - reading it twice, perhaps this is the pagination logic? If so, I would add some comments just for our future ourselves to understand 😅

Reply:
This is us using the pagination of the dockerhub API, yes - requesting the next page if `response.next` is defined. I'll add a comment to that effect.